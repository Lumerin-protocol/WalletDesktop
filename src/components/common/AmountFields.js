import React from 'react';

import { FieldBtn } from './Btn';
import TextInput from './TextInput';
import TxIcon from '../icons/TxIcon';
import Flex from './Flex';
import Sp from './Spacing';

export default function AmountFields({
  coinPlaceholder,
  usdPlaceholder,
  coinSymbol,
  onMaxClick,
  coinAmount,
  usdAmount,
  autoFocus,
  onChange,
  errors
}) {
  return (
    <Flex.Row justify="space-between">
      <Flex.Item grow="1" basis="0">
        <FieldBtn
          data-testid="max-btn"
          tabIndex="-1"
          onClick={onMaxClick}
          float
        >
          MAX
        </FieldBtn>
        <TextInput
          placeholder={coinPlaceholder}
          data-testid="coinAmount-field"
          autoFocus={autoFocus}
          onChange={onChange}
          error={errors.coinAmount}
          value={coinAmount}
          label={`Amount (${coinSymbol})`}
          id="coinAmount"
        />
      </Flex.Item>
      <Sp mt={6} mx={1}>
        <TxIcon />
      </Sp>
      <Flex.Item grow="1" basis="0">
        <TextInput
          placeholder={usdPlaceholder}
          data-testid="usdAmount-field"
          onChange={onChange}
          error={errors.usdAmount}
          value={usdAmount}
          label="Amount (USD)"
          id="usdAmount"
        />
      </Flex.Item>
    </Flex.Row>
  );
}
