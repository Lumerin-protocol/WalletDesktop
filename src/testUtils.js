/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
