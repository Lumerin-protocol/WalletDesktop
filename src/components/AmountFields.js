import { TextInput, FieldBtn, TxIcon, Flex, Sp } from './common'
import * as utils from '../utils'
import PropTypes from 'prop-types'
import React from 'react'
import Web3 from 'web3'

export default class AmountFields extends React.Component {
  static propTypes = {
    availableETH: PropTypes.string.isRequired,
    ethAmount: PropTypes.string,
    usdAmount: PropTypes.string,
    autoFocus: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    errors: PropTypes.shape({
      ethAmount: PropTypes.string,
      usdAmount: PropTypes.string
    }).isRequired
  }

  static SMALL_VALUE_PLACEHOLDER = '< 0.01'
  static INVALID_PLACEHOLDER = 'Invalid amount'

  static initialState = {
    ethAmount: null,
    usdAmount: null
  }

  static onInputChange = (state, ETHprice, id, value) => ({
    ...state,
    usdAmount:
      id === 'ethAmount'
        ? utils.toUSD(
            value,
            ETHprice,
            AmountFields.INVALID_PLACEHOLDER,
            AmountFields.SMALL_VALUE_PLACEHOLDER
          )
        : state.usdAmount,
    ethAmount:
      id === 'usdAmount'
        ? utils.toETH(value, ETHprice, AmountFields.INVALID_PLACEHOLDER)
        : state.ethAmount
  })

  onMaxClick = () => {
    const ethAmount = Web3.utils.fromWei(this.props.availableETH)
    this.props.onChange({ target: { id: 'ethAmount', value: ethAmount } })
  }

  render() {
    const { autoFocus, ethAmount, usdAmount, onChange, errors } = this.props

    return (
      <Flex.Row justify="space-between">
        <Flex.Item grow="1" basis="0">
          <FieldBtn
            data-testid="max-btn"
            tabIndex="-1"
            onClick={this.onMaxClick}
            float
          >
            MAX
          </FieldBtn>
          <TextInput
            data-testid="ethAmount-field"
            placeholder={
              ethAmount === AmountFields.INVALID_PLACEHOLDER
                ? AmountFields.INVALID_PLACEHOLDER
                : '0.00'
            }
            autoFocus={autoFocus}
            onChange={onChange}
            error={errors.ethAmount}
            label="Amount (ETH)"
            value={
              ethAmount === AmountFields.INVALID_PLACEHOLDER ? '' : ethAmount
            }
            id="ethAmount"
          />
        </Flex.Item>
        <Sp mt={6} mx={1}>
          <TxIcon />
        </Sp>
        <Flex.Item grow="1" basis="0">
          <TextInput
            data-testid="usdAmount-field"
            placeholder={
              usdAmount === AmountFields.INVALID_PLACEHOLDER
                ? AmountFields.INVALID_PLACEHOLDER
                : usdAmount === AmountFields.SMALL_VALUE_PLACEHOLDER
                  ? AmountFields.SMALL_VALUE_PLACEHOLDER
                  : '0.00'
            }
            onChange={onChange}
            error={errors.usdAmount}
            label="Amount (USD)"
            value={
              usdAmount === AmountFields.INVALID_PLACEHOLDER ||
              usdAmount === AmountFields.SMALL_VALUE_PLACEHOLDER
                ? ''
                : usdAmount
            }
            id="usdAmount"
          />
        </Flex.Item>
      </Flex.Row>
    )
  }
}
