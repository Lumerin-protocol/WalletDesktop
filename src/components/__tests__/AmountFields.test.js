import { Simulate } from 'react-testing-library'
import * as testUtils from '../../testUtils'
import AmountFields from '../common/AmountFields'
import React from 'react'

const element = (
  <AmountFields
    usdPlaceholder="0.00"
    ethPlaceholder="0.00"
    availableETH="100"
    onMaxClick={jest.fn()}
    onChange={jest.fn()}
    errors={{}}
  />
)

describe('<AmountFields/>', () => {
  it('should match its snapshot', () => {
    const { container } = testUtils.reduxRender(element)
    expect(container).toMatchSnapshot()
  })
})

export function runEditTests(el, initialState, rate) {
  it('updates the USD field when ETH field changes', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    expect(ethField.value).toBe('')
    expect(usdField.value).toBe('')
    ethField.value = '1'
    Simulate.change(ethField)
    expect(usdField.value).toBe(rate.toString())
  })

  it.skip('updates the ETH field when USD field changes', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    usdField.value = '500'
    Simulate.change(usdField)
    expect(ethField.value).toBe((500 / rate).toString())
  })

  it.skip('updates ETH and USD fields when MAX button is clicked', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    Simulate.click(getByTestId('max-btn'))
    expect(ethField.value).toBe('5000')
    expect(usdField.value).toBe((5000 * rate).toString())
  })

  it.skip('displays a "Invalid amount" placeholder in USD field when ETH value is invalid', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    expect(usdField.placeholder).toBe('0.00')
    ethField.value = 'foo'
    Simulate.change(ethField)
    expect(usdField.value).toBe('')
    expect(usdField.placeholder).toBe('Invalid amount')
  })

  it.skip('displays a "< $0.01" placeholder in USD field when ETH equivalent value is small', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    expect(usdField.placeholder).toBe('0.00')
    ethField.value = '0.0000001'
    Simulate.change(ethField)
    expect(usdField.value).toBe('')
    expect(usdField.placeholder).toBe('< 0.01')
  })

  it.skip('displays a "Invalid amount" placeholder in ETH field when USD value is invalid', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState)
    const ethField = getByTestId('ethAmount-field')
    const usdField = getByTestId('usdAmount-field')
    expect(ethField.placeholder).toBe('0.00')
    usdField.value = 'foo'
    Simulate.change(usdField)
    expect(ethField.value).toBe('')
    expect(ethField.placeholder).toBe('Invalid amount')
  })
}

// eslint-disable-next-line max-params
export function runValidateTests(
  el,
  initialState,
  formTestId,
  fieldTestId = 'ethAmount-field'
) {
  it('displays an error if AMOUNT is not provided', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { [fieldTestId]: '' },
      errors: { [fieldTestId]: 'Amount is required' }
    })
  })

  it('displays an error if AMOUNT is negative', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { [fieldTestId]: '-1' },
      errors: { [fieldTestId]: 'Amount must be greater than 0' }
    })
  })

  it('displays an error if AMOUNT is an invalid value', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { [fieldTestId]: 'foo' },
      errors: { [fieldTestId]: 'Invalid amount' }
    })
  })

  it('displays an error if AMOUNT is more than what we have', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { [fieldTestId]: '10000' },
      errors: { [fieldTestId]: 'Insufficient funds' }
    })
  })

  it('displays an error if AMOUNT is not weiable', () => {
    const { getByTestId } = testUtils.reduxRender(el, initialState)
    testUtils.testValidation(getByTestId, formTestId, {
      formData: { [fieldTestId]: '0.0000000000000000000000000000001' },
      errors: { [fieldTestId]: 'Invalid amount' }
    })
  })
}
