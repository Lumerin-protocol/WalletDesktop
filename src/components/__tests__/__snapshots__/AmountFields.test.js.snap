// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountFields/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-htoDjs hAEOmi"
  >
    <div
      class="sc-gzVnrw bHqtcW"
    >
      <button
        class="sc-htpNat fpAZyb"
        data-testid="max-btn"
        tabindex="-1"
        type="button"
      >
        MAX
      </button>
      <div>
        <label
          class="sc-bxivhb htFESG"
          for="ethAmount"
        >
          Amount (ETH)
        </label>
        <input
          class="sc-ifAKCX kSbDeA"
          data-testid="ethAmount-field"
          id="ethAmount"
          placeholder="0.00"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-iwsKbI fGVAip"
    >
      <svg
        class="sc-bZQynM hKkckV"
        viewBox="0 0 24 24"
      >
        <path
          d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
        />
      </svg>
    </div>
    <div
      class="sc-gzVnrw bHqtcW"
    >
      <div>
        <label
          class="sc-bxivhb htFESG"
          for="usdAmount"
        >
          Amount (USD)
        </label>
        <input
          class="sc-ifAKCX kSbDeA"
          data-testid="usdAmount-field"
          id="usdAmount"
          placeholder="0.00"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
